### PR TITLE
Use Scala code to decode PGP_SECRET

### LINKED ISSUE #40: Use Scala code instead of shelling out to echo and base64?
Hi @ckipp01,

I'm just trying to use this plugin on Windows where `echo` and `base64` are not natively available.
Would it be possible to use native Scala code for that?
Is using third-party libraries an option or would you be concerned about the necessity to download additional stuff?

Thank you for creating this!

### PR DESCRIPTION
**Pull request by nightscape**
Fixes #40

### CODE REVIEW

File: plugin/src/io/kipp/mill/ci/release/CiReleaseModule.scala [hunk 1]
```diff
@@ -159,13 +161,13 @@ object ReleaseModule extends ExternalModule {
         "Unable to call gpg. Are you sure it's installed and set up correctly?"
       )
     } else {
-      val echo = os.proc("echo", envTask().pgpSecret).spawn()
-      val decoded = os.proc("base64", "--decode").spawn(stdin = echo.stdout)
+      val pgpSecret = envTask().pgpSecret
+      val decoded = new String(Base64.getDecoder.decode(pgpSecret.getBytes(StandardCharsets.UTF_8)))
```
Comment (ckipp01) [inline plugin/src/io/kipp/mill/ci/release/CiReleaseModule.scala:165]: So I think if we're going to move into Scala-land here we should probably start also catching stuff that can throw. We should probably wrap this in a try here to catch the exception that could happen in `decode` if it's not a valid base64 encoded string.

**Review by ckipp01 — CHANGES_REQUESTED** [review]
Thanks @nightscape! So even on windows though, won't you still have issues with some of the other methods being used like `gpg`, or will that actually be available to you?

**Comment by nightscape** [comment]
I have `gpg` available. Not sure atm if it was available by default, or if I installed it myself.

File: plugin/src/io/kipp/mill/ci/release/CiReleaseModule.scala (hunk 1 above)
Comment (ckipp01) [reply plugin/src/io/kipp/mill/ci/release/CiReleaseModule.scala]: What I'd recommend is wrapping the entire thing in a try and catch the `IllegalArgumentException`. Something like this:

```scala
      try {
        val decoded = new String(
          Base64.getDecoder.decode(pgpSecret.getBytes(StandardCharsets.UTF_8))
        )

        // https://dev.gnupg.org/T2313
        val imported = os
          .proc("gpg", "--batch", "--import", "--no-tty")
          .call(stdin = decoded)

        if (imported.exitCode != 0)
          Result.Failure(
            "Unable to import your pgp key. Make sure your secret is correct."
          )
      } catch {
        case e: IllegalArgumentException =>
          Result.Failure(
            s"Invalid secrete, unable to decode it: ${e.getMessage()}"
          )
        case e: Throwable => Result.Failure(e.getMessage())
      }
```

Comment (nightscape) [reply plugin/src/io/kipp/mill/ci/release/CiReleaseModule.scala]: Looks good!
Should the `case e: Throwable` be a `case e: Exception`?
Afair, one should not catch `Throwable`.

**Comment by nightscape** [comment]
@ckipp01 how about the new version?
Not sure if you like the `Either`-based for-comprehension?

File: plugin/src/io/kipp/mill/ci/release/CiReleaseModule.scala (hunk 1 above)
Comment (ckipp01) [reply plugin/src/io/kipp/mill/ci/release/CiReleaseModule.scala]: Correct, this was quickly pseudo-coded. As an alternative we can probably catch `NonFatal(e)` here.

**Comment by nightscape** [comment]
I just added code that removes whitespaces from the `PGP_SECRET`, otherwise the Base64 decoder breaks e.g. on new-lines.

File: plugin/src/io/kipp/mill/ci/release/CiReleaseModule.scala [hunk 2]
```diff
@@ -153,25 +157,27 @@ object ReleaseModule extends ExternalModule {
     */
   private def setupGpg() = T.task {
     T.log.info("Attempting to setup gpg")
-    val versionCall = os.proc("gpg", "--version").call()
-    if (versionCall.exitCode != 0) {
-      Result.Failure(
-        "Unable to call gpg. Are you sure it's installed and set up correctly?"
+    val pgpSecret = envTask().pgpSecret.replaceAll("\\s", "")
```
Comment (ckipp01) [inline plugin/src/io/kipp/mill/ci/release/CiReleaseModule.scala:160]: Would trim do the same thing here?

Comment (nightscape) [reply plugin/src/io/kipp/mill/ci/release/CiReleaseModule.scala:160]: Just tested this in Ammonite, and `trim` only seems to remove leading and trailing spaces:
```scala
" foo bar baz\t".trim // => "foo bar baz"
```
PGP_SECRETS often consist of multiple lines which would not be covered by `trim`.

**Comment by nightscape** [comment]
@ckipp01 this is good to merge from my side 😃

**Review by ckipp01 — APPROVED** [review]
Hey @nightscape, sorry for the delay on this. Things got a bit hectic and I've put it off. Let's give it a go. If something seems off, we can just revert. Thanks again for your work on this.

**Pull request merged**